### PR TITLE
Fix SpaceAwareAccessor based on usage experiment in View

### PR DESCRIPTION
**Pull request by crtrott**
- fix offset type of SpaceAwareAccesor
- Improved test to catch the mistake, and added a typedef for nested_accessor which I believe every wrapping accessor should have
- fix convertibility properties to match View convertibility

### CODE REVIEW

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp [hunk 1]
```diff
@@ -81,8 +83,7 @@ struct SpaceAwareAccessor {
   }
 
   KOKKOS_FUNCTION
-  constexpr data_handle_type offset(data_handle_type p, size_t i) const
-      noexcept {
+  constexpr auto offset(data_handle_type p, size_t i) const noexcept {
```
Comment (masterleinad) [inline core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:86]: The return type should really be `typename offset_policy​::​data_handle_type` (according to https://eel.is/c++draft/mdspan.accessor.reqmts#9), right?

**Review by nliber — APPROVED** [review]

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:86]: jup but that is the same as nested_acc.offset return type

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp [hunk 2]
```diff
@@ -34,35 +34,30 @@ namespace Impl {
...
-  template <class OtherNestedAccessorType,
-            std::enable_if_t<std::is_constructible_v<NestedAccessor,
-                                                     OtherNestedAccessorType>,
-                             int> = 0>
+  template <
+      class OtherMemorySpace, class OtherNestedAccessorType,
+      std::enable_if_t<
+          MemorySpaceAccess<MemorySpace, OtherMemorySpace>::assignable &&
+              std::is_constructible_v<NestedAccessor, OtherNestedAccessorType>,
+          int> = 0>
```
Comment (dalg24) [inline core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:58]: This change does not seem to belong to this PR which the description advertise as being about fixing the return type of the `SpaceAwareAccessor::offset` member function.
Please withdraw it or make the case that it belongs here and update the title and description.

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp [hunk 3]
```diff
@@ -34,35 +34,30 @@ namespace Impl {
...
   using reference        = typename NestedAccessor::reference;
   using data_handle_type = typename NestedAccessor::data_handle_type;
-
   using offset_policy =
       SpaceAwareAccessor<MemorySpace, typename NestedAccessor::offset_policy>;
 
-  using memory_space = MemorySpace;
+  // Specific to SpaceAwareAccessor
+  using memory_space         = MemorySpace;
+  using nested_accessor_type = NestedAccessor;
```
Comment (dalg24) [inline core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:46]: I don't necessary have an issue with adding that public member type but you don't say we might want to have it and it is not used.

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp (hunk 1 above)
Comment (masterleinad) [reply core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:79]: I would prefer spelling out the return type here instead of using `auto`.

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp (hunk 2 above)
Comment (crtrott) [reply core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:58]: I updated title and description

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:79]: done

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp (hunk 3 above)
Comment (crtrott) [reply core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:46]: I still think we should have it, and I did add the nested_accessor() function as in https://eel.is/c++draft/linalg.scaled

**Review by masterleinad — APPROVED** [review]

File: core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp (hunk 3 above)
Comment (dalg24) [reply core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:46]: `std::linalg::scaled_accessor` does not define a member type for it.  Anyway, I don't feel too strongly about it.

Comment (crtrott) [reply core/src/View/MDSpan/Kokkos_MDSpan_Accessor.hpp:46]: Yeah I know, the comma in my reply was basically saying "I do think we should have the typedef" and orthogonal to that "I also added the nested_accessor() function as in scaled_accessor"
